### DISCUSSION TITLE
make dev fails on fresh Debian VM due to permission issues

### DISCUSSION THREAD

**Discussion by kaysond** [immich-app/immich#10788] [Q&A]
### The bug

`make dev` fails on a fresh OS install and fresh clone of the latest release.
![image](https://github.com/immich-app/immich/assets/1147328/1e007e32-4b18-4bf5-8848-e2041d26f47b)

Doing a `chmod a+w ./web` then running again fixes that problem, but runs into another:
![image](https://github.com/immich-app/immich/assets/1147328/db221bfa-2e15-4a8d-89aa-ebc5e821b232)

This directory doesn't get created until somewhat into the build process, so you could keep doing chmods, but ultimately the issue is caused by using the `node` user in https://github.com/immich-app/immich/blob/main/web/Dockerfile
Removing those bits and running the container as root works. I'm guessing many people are logged in with the default uid of 1000, which is what the node user in the container uses. Thus permissions just work. But if you're not using that uid, it won't work.

### The OS that Immich Server is running on

Debian 12

### Version of Immich Server

v1.93.2

### Version of Immich Mobile App

N/A

### Platform with the issue

- [ ] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
N/A
```

### Your .env content

```Shell
docker/sample.env
```

### Reproduction steps

```bash
1. Install Debian 12 and set up a root user only
2. Install Docker
3. Clone immich release
4. Run `make dev`
```

**Comment by jrasm91** [comment]
The web container runs as non-root by default. I believe it uses user 1000. If your system doesn't have that user or you are running as root, you should configure docker compose to run that container as your desired user instead.

**Reply by kaysond** [reply]
Right. You've got

https://github.com/immich-app/immich/blob/96f29cefeb35608e9f16262d1897303a5246c0dc/web/Dockerfile#L4

and in the base node image [Dockerfile](https://github.com/nodejs/docker-node/blob/58d00e1735812b2a237b80785ee0924f061cf728/22/alpine3.19/Dockerfile#L5-L6), there is

```dockerfile
RUN addgroup -g 1000 node \
    && adduser -u 1000 -G node -s /bin/sh -D node \
    && apk add --no-cache \
```

The problem is that you then mount the contents of `./web` into the container:

https://github.com/immich-app/immich/blob/96f29cefeb35608e9f16262d1897303a5246c0dc/docker/docker-compose.dev.yml#L62

So if uid `1000` doesn't exist on the host, it can't access those files period.

Yes, you can fix it, but that means either modifying `web/Dockerfile` to change the user in the image, or modifying the `docker/docker-compose.dev.yml` to choose a different user at runtime. It's just inconvenient and probably not a great thing to carry around. Ideally, the dev setup wouldn't require you to have any particular UID/user configured on your host.

I'm not sure if/why the web container needs to be running as uid 1000, but I think you could just run it as root and everything would work out. It would even work for rootless docker.

**Reply by jrasm91** [reply]
If you run it as root, then everybody else (who runs on a "normal" system) now has permission issues since the files created in the container have root permissions on the host.

**Reply by kaysond** [reply]
Are there any files created in the container that are persisted? I thought they all go in the docker volume (not the bind mount). If you run as non-root it also causes permission issues for people running rootless because the 1000 uid gets mapped to something else. There are lots of reasons to *not* force a container to use a certain uid. 

If no one else is complaining about this and you're not inclined to fix it, feel free to close.

**Reply by jrasm91** [reply]
I'm pretty sure the `.svelte-kit/` folder is created by the web process, running inside the container.

**Reply by jrasm91** [reply]
We used to run this container as root and we switched it to non-root. I don't think I've heard any problem about it _not_ being root, besides this issue. I also assume running it as non-root is also just a general best practice. If it doesn't need to run as root, why should it?

**Reply by kaysond** [reply]
I'm not sure it matters much for something like a dev container that isn't exposed outside the host, but I agree - might as well not run as root. But the problem isn't that; it's that the container is being forced to run as a particular uid.

I think the "docker" way to do this for development is by running the daemon rootless, so uid 0 in the container maps to your uid on the host. Another option, which is what the LSIO containers (and immich too, I think) do is to start as root and drop to an unprivileged user based on env vars. Maybe less practical with the way the make script and docker compose are set up. 

It's also worth pointing out that there's a risk in forcing any uid in that the logged in user may not be the same one, so you don't really know who you're running as (in the sense that I think this is similar to running as root).

**Reply by jrasm91** [reply]
I'm open to changing the setup as long as it works out of the box in both situations

**Reply by kaysond** [reply]
Sounds good. I'll take a look at it when I get a chance.
